arch/risc-v/src/common/espressif/esp_wifi_utils.c

```diff
@@ -0,0 +1,643 @@
+        {
+          /* Scan specific bssid */
+
+          memcpy(target_mac, req->bssid.sa_data, MAC_LEN);
+          config->bssid = &target_mac[0];
+        }
+    }
+  else
+    {
+      /* Default scan parameters */
+
+      wlinfo("INFO: Use default scan parameters\n");
+      config->scan_type = WIFI_SCAN_TYPE_ACTIVE; /* Active scan */
+    }
+
+  esp_wifi_start();
+  ret = esp_wifi_scan_start(config, false);
+  if (ret != OK)
+    {
+      wlerr("ERROR: Scan error, ret: %d\n", ret);
+    }
+  else
+    {
+      /* Allocate buffer to store scan result */
+
+      if (priv->scan_result == NULL)
+        {
+          priv->scan_result = kmm_malloc(WIFI_SCAN_RESULT_SIZE);
+          if (priv->scan_result == NULL)
+            {
+              wlerr("ERROR: Cannot allocate result buffer\n");
+              ret = -ENOMEM;
+            }
+          else
+            {
+              memset(priv->scan_result, 0x0, WIFI_SCAN_RESULT_SIZE);
+            }
+        }
+    }
+
+  if (config)
+    {
+      kmm_free(config);
+      config = NULL;
+      wlinfo("INFO: start scan\n");
+    }
+
+  g_scan_priv.scan_status = ESP_SCAN_RUN;
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: esp_wifi_get_scan_results
+ *
+ * Description:
+ *   Get scan result
+ *
+ * Input Parameters:
+ *   iwr - The argument of the ioctl cmd
+ *
+ * Returned Value:
+ *   OK on success (positive non-zero values are cmd-specific)
+ *   Negated errno returned on failure.
+ *
+ ****************************************************************************/
+
+int esp_wifi_get_scan_results(struct iwreq *iwr)
+{
+  int ret = OK;
+  static bool scan_block = false;
+  struct wifi_scan_result *priv = &g_scan_priv;
+
+  if (g_scan_priv.scan_status == ESP_SCAN_RUN)
+    {
+      irqstate_t irqstate = enter_critical_section();
+      if (scan_block == false)
+        {
+          scan_block = true;
+          leave_critical_section(irqstate);
+          nxsem_tickwait(&priv->scan_signal, SEC2TICK(SCAN_TIME_SEC));
+          scan_block = false;
+        }
+      else
+        {
+          leave_critical_section(irqstate);
+          ret = -EINVAL;
+          goto exit_failed;
+        }
+    }
+  else if (g_scan_priv.scan_status == ESP_SCAN_DISABLED)
+    {
+      return -EINVAL;
+    }
+
+  if ((iwr == NULL) || (g_scan_priv.scan_status != ESP_SCAN_DONE))
+    {
+      ret = -EINVAL;
+      goto exit_failed;
+    }
+
+  if (priv->scan_result == NULL)
+    {
+      /* Result have already been requested */
+
+      ret = OK;
+      iwr->u.data.length = 0;
+      goto exit_failed;
+    }
+
+  if (iwr->u.data.pointer == NULL ||
+      iwr->u.data.length < priv->scan_result_size)
+    {
+      /* Stat request, return scan_result_size */
+
+      ret = -E2BIG;
+      iwr->u.data.pointer = NULL;
+      iwr->u.data.length = priv->scan_result_size;
+      goto exit_failed;
+    }
+
+  if (priv->scan_result_size <= 0)
```

Comment:
Maybe this should be checked after (together with) `if (priv->scan_result == NULL)` above before we do `iwr->u.data.length < priv->scan_result_size` at line 305?